File: clojure/nucleotide-count/src/nucleotide_count.clj
(ns nucleotide-count)

(def valid-nucleotides #{\A \C \G \T})

(defn count-of-nucleotide-in-strand [nucleotide strand]
  (if (valid-nucleotides nucleotide)
    (count (filter #(= % nucleotide) strand))
    (throw (Exception. "invalid base"))))

(defn nucleotide-counts [strand]
  (zipmap valid-nucleotides
          (map #(count-of-nucleotide-in-strand % strand) valid-nucleotides)))

File: julia/nucleotide-count/nucleotide-count.jl
"""
    count_nucleotides(strand)

The count of each nucleotide within `strand` as a dictionary.

Invalid strands raise a `DomainError`.

"""
function count_nucleotides(strand)
    nucleotides = Dict('A' => 0, 'C' => 0, 'G' => 0, 'T' => 0)
    for base in strand
        if haskey(nucleotides, base)
            nucleotides[base] += 1
        else
            throw(DomainError(base, "invalid base"))
        end
    end
    nucleotides
end
